Source/WebKit/NetworkProcess/Downloads/Download.cpp

```diff
@@ -141,6 +141,13 @@ void Download::didFinish()
 {
     DOWNLOAD_RELEASE_LOG("didFinish: (id = %" PRIu64 ")", downloadID().toUInt64());
 
+#if HAVE(MODERN_DOWNLOADPROGRESS)
+    // If we are using a placeholder, we will call didFinish in Download::setFinalURL.
+    // This is because m_downloadManager->downloadFinished will delete the Download object.
+    if (m_isUsingPlaceholder)
```

Comment:
This split flow isn't great.  It looks like we don't revoke the sandbox extension in the setFinalURL flow.  Do we know that setFinalURL will always be called after didFinish if this is true, or could it be called before?

Reply:
We will still revoke the sandbox extension, since this is done in Download::didFinish(), which we call from Download::setFinalURL. Download::setFinalURL should always be called after didFinish, since it is based on the progress we report.